File: 121-Best Time to Buy and Sell Stock/best_time_buy_sell_stock.dart
import 'dart:math';

void main() {
  final prices = [7, 1, 5, 3, 6, 4];
  print(BestTimeBuySellStock.maxProfit(prices).toString());
}

abstract class BestTimeBuySellStock {
  static int maxProfit(List<int> prices) {
    //
    // First attempt 27 June 2024
    // Runtime: 410ms, Memory: 189.11MB
    // Time complexity: O(n^2), Space complexity: O(1)
    //
    int profit = 0;
    int buyPrice = prices[0];

    for (int i = 1; i < prices.length; i++) {
      if (prices[i] > buyPrice) {
        profit = max(profit, prices[i] - buyPrice);
      } else {
        buyPrice = prices[i];
      }
    }

    return profit;
  }
}

File: 121-Best Time to Buy and Sell Stock/BestTimeBuySellStock.kt
package main

fun main() {
    val prices = intArrayOf(7, 1, 5, 3, 6, 4)
    println(BestTimeBuySellStock.maxProfit(prices))
}

object BestTimeBuySellStock {
    fun maxProfit(prices: IntArray): Int {
        //
        // First attempt 27 June 2024 15:45
        // Runtime: 467ms, Memory: 55.70MB
        // Time complexity: O(n), Space complexity: O(1)
        //
        var profit = 0
        var buyPrice = prices[0]

        for (i in 1..<prices.size) {
            if (prices[i] > buyPrice) {
                profit = maxOf(profit, prices[i] - buyPrice)
            } else {
                buyPrice = prices[i]
            }
        }

        return profit
    }
}

